### PR TITLE
Issue 61 Fix with old style "Access" instead of "Rental"

### PR DESCRIPTION
**Pull request by eemhu**
Add new phaser implementation rather than using a custom count and lock solution to BufferLease and Access.

### CODE REVIEW

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java [hunk 1]
```diff
@@ -1,127 +1,100 @@
...
     @Override
     public void addRef() {
-        lock.lock();
-        try {
-            refCount++;
-        }
-        finally {
-            lock.unlock();
+        int a = this.phaser.register();
+        if (a < 0) {
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java:43]: Might as well be inline as this variable is not used anywhere?
`if (this.phaser.register() < 0)`

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java [hunk 2]
```diff
@@ -1,127 +1,100 @@
...
-        return false;
+    public synchronized boolean attemptRelease() {
+        /*
+        int a = phaser.arriveAndDeregister();
+        if (a < 0) {
+            throw new IllegalStateException("poks");
+        }
+        return phaser.isTerminated();
+
+         */
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java:75]: Remove commented out code, doesn't look like something that should stay here for future reference

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java [hunk 3]
```diff
@@ -38,30 +39,43 @@ public BufferLeasePool() {
...
+        } else {
+            // otherwise, wrap bufferContainer with phaser decorator (bufferLease)
+            bufferLease = new BufferLeaseImpl(bufferContainer, this);
         }
 
-        bufferLease.addRef(); // all start with one ref
-
         if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug("returning bufferLease id <{}> with refs <{}> at buffer position <{}>", bufferLease.id(), bufferLease.refs(), bufferLease.buffer().position());
+            LOGGER.debug("returning bufferLease id <{}> with refs<{}> at buffer position <{}>", bufferLease.id(), bufferLease.refs(), bufferLease.buffer().position());
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java:64]: Did `refs<{}>` lose space intentionally?

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java [hunk 4]
```diff
@@ -38,30 +39,43 @@ public BufferLeasePool() {
...
+        }
+
+        if (bufferLease.buffer().position() != 0) {
+            LOGGER.error("dirty buffer in pool, terminating");
+            System.exit(1);
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java:69]: Is hard exiting really wanted here? If so, I'd use enum values for exit like `System.exit(ExitCodes.DIRTY_BUFFER_IN_POOL);` to track what are defined and to catch deal with return codes in shell and so on

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java [hunk 5]
```diff
@@ -1,79 +1,44 @@
...
-
-    private long accessCount;
-    private boolean terminated;
-    private final Lock lock;
+    final Phaser phaser;
     public Access() {
-        this.accessCount = 0; // TODO consider using a semaphore
-        this.terminated = false;
-        this.lock = new ReentrantLock();
+        this.phaser = new Phaser(1);
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java:9]: Add a note what the argument `1` is for as it is not immediately obvious

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java [hunk 6]
```diff
@@ -1,79 +1,44 @@
...
+            throw new IllegalStateException("Cannot release an open lease");
+        } else if (phaser.isTerminated()) {
+            throw new IllegalStateException("Phaser was already terminated");
         }
+        // deregister released lease
+        phaser.arriveAndDeregister();
+    }
+
+    public boolean terminated() {
+        return phaser.isTerminated();
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java:42]: Should that be `isTerminated()`?

**Comment by StrongestNumber9** [comment]
General note: Noticed mixed use of `this.classVariable` and `classVariable` but didn't comment on all of them. Use them consistently in one way or another

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java [hunk 7]
```diff
@@ -78,23 +92,24 @@ public List<BufferLease> take(long size) {
...
             while (queue.peek() != null) {
                 if (lock.tryLock()) {
                     while (true) {
-                        BufferLease queuedBufferLease = queue.poll();
-                        if (queuedBufferLease == null) {
+                        BufferContainer queuedBufferContainer = queue.poll();
+                        if (queuedBufferContainer == null) {
                             break;
                         }
                     }
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java:115]: Will this implementation consume all available processing power in a busy loop until poll returns null?

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java [hunk 8]
```diff
@@ -1,28 +1,31 @@
...
             access.release(this);
+        } else {
+            throw new IllegalStateException("lease not open, can't close");
         }
     }
 
     public boolean isOpen() {
-        return isOpen;
+        return !subPhaser.isTerminated();
     }
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java:30]: I think `isTerminated()` should be used instead of inversing returns like it was used few lines above

**Comment by eemhu** [comment]
added fixes till "wrong classname" comment

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java [hunk 9]
```diff
@@ -1,28 +1,31 @@
...
+        if (subPhaser.isTerminated()) {
+            throw new IllegalStateException("lease phaser was terminated, can't close");
         }
-        else {
-            isOpen = false;
+        else if (isOpen()) {
+            subPhaser.arriveAndDeregister(); // should terminate
             access.release(this);
+        } else {
+            throw new IllegalStateException("lease not open, can't close");
```
Comment (StrongestNumber9) [inline src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java:24]: Can this code even be reached?

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java (hunk 1 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java]: fix [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java [hunk 10]
```diff
@@ -1,127 +1,100 @@
...
-                throw new IllegalStateException("refs must not be negative");
-            }
-
-            refCount = newRefs;
-        }
-        finally {
-            lock.unlock();
+        int a = phaser.arriveAndDeregister();
+        if (a < 0) {
+            throw new IllegalStateException("Cannot remove reference, BufferLease phaser was already terminated!");
```
[the remark this answers is not in the document]
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java]: fix [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java (hunk 2 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeaseImpl.java]: fix [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java (hunk 3 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java]: fix [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java (hunk 4 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java]: fix: Add IllegalStateException [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java (hunk 5 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java:10]: add comment fix [c447182](https://github.com/teragrep/rlp_03/pull/84/commits/c44718257663e780cb7e9699b9c1159fcf9517e9)

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java (hunk 8 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java:28]: changed to isTerminated and changed all usages to be negated: isOpen -> !isTerminated.

**Comment by eemhu** [comment]
added fixes to rest of the comments.

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java [hunk 11]
```diff
@@ -38,23 +39,35 @@ public BufferLeasePool() {
...
-        BufferLease bufferLease = queue.poll();
-        if (bufferLease == null) {
-            bufferLease = new BufferLeaseImpl(bufferId.incrementAndGet(), byteBufferSupplier.get());
+        BufferContainer bufferContainer = queue.poll();
+        BufferLease bufferLease;
+        if (bufferContainer == null) {
+            // if queue is empty or stub object, create a new BufferContainer and BufferLease.
+            bufferLease = new BufferLeaseImpl(
+                    new BufferContainerImpl(bufferId.incrementAndGet(), byteBufferSupplier.get())
+                    ,
```
Comment (kortemik) [inline src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java:56]: auto-indent

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java [hunk 12]
```diff
@@ -1,79 +1,45 @@
...
 
-    public boolean terminated() {
-        lock.lock();
-        try {
-            return terminated;
-        }
-        finally {
-            lock.unlock();
+    public void terminate() throws IllegalStateException {
+        phaser.arriveAndDeregister(); // registered=0, should terminate phaser
```
Comment (kortemik) [inline src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java:25]: i think these should rather check the return value of phaser.arriveAndDeregister(); being non-negative, because terminated phaser returns negative ones

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java [hunk 13]
```diff
@@ -1,79 +1,45 @@
 package com.teragrep.rlp_03.context.frame.access;
 
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.Phaser;
 import java.util.function.Supplier;
 
 public final class Access implements Supplier<Lease> {
```
Comment (kortemik) [inline src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java:6]: I recommend moving the whole Access stuff to another pr because it needs more work.

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java [hunk 14]
```diff
@@ -1,28 +1,29 @@
...
 public class Lease implements AutoCloseable {
 
     private final Access access;
-
-    private volatile boolean isOpen;
+    private final Phaser subPhaser;
     Lease(Access access) {
         this.access = access;
-        this.isOpen = true;
+        this.subPhaser = new Phaser(1);
```
Comment (kortemik) [inline src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java:11]: these should be of the Access's phaser, so when sub-phaser terminates, so terminates the one in Access

**Review by kortemik — CHANGES_REQUESTED** [review]
move Access stuff to it's own pr and this will do

**Comment by eemhu** [comment]
@kortemik Reverted to main branch Access / Lease / AccessTest. This PR should now only contain the changed BufferLease, will move Access phaser to new pull request.

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLease.java [hunk 15]
```diff
@@ -3,9 +3,12 @@
 import java.nio.ByteBuffer;
 
 public interface BufferLease {
+    BufferContainer bufferContainer();
```
[the remark this answers is not in the document]
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/buffer/BufferLease.java]: removed https://github.com/teragrep/rlp_03/pull/84/commits/cb60de719532b616c8ca056328011f9f6197fa5d

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java (hunk 11 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/buffer/BufferLeasePool.java]: fix indent https://github.com/teragrep/rlp_03/pull/84/commits/cb60de719532b616c8ca056328011f9f6197fa5d

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java (hunk 13 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/frame/access/Access.java]: will move

File: src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java (hunk 14 above)
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/frame/access/Lease.java]: will change to access pr

**Review by kortemik — APPROVED** [review]

**Review by StrongestNumber9 — APPROVED** [review]
LGTM

File: src/main/java/com/teragrep/rlp_03/context/buffer/BufferContainerImpl.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    public synchronized ByteBuffer buffer() {
+        return buffer;
+    }
+
+    @Override
+    public String toString() {
+        return "BufferLease{" +
+                "buffer=" + buffer +
+                '}';
+    }
```
[the remark this answers is not in the document]
Comment (eemhu) [reply src/main/java/com/teragrep/rlp_03/context/buffer/BufferContainerImpl.java]: added id and also changed from BufferLease to BufferContainer.
